src/main/java/com/mongodb/jdbc/utils/X509Authentication.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.jdbc.utils;
+
+import com.mongodb.jdbc.logging.MongoLogger;
+import java.io.FileReader;
+import java.security.*;
+import java.security.cert.Certificate;
+import java.util.logging.Level;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.bouncycastle.pkcs.PKCS8EncryptedPrivateKeyInfo;
+import org.bouncycastle.pkcs.jcajce.JcePKCSPBEInputDecryptorProviderBuilder;
+
+public class X509Authentication {
+    private static final String BOUNCY_CASTLE = "BC";
+    private final MongoLogger logger;
+
+    static {
+        if (Security.getProvider(BOUNCY_CASTLE) == null) {
+            Security.addProvider(new BouncyCastleProvider());
```

Comment:
This makes me quite nervous, as it effects global application state.
Is there an alternative to adding a global provider like this?

Reply:
Thanks for pointing this out.  I will try finding an alternative to adding the global provider and let you know.

Reply:
I made a change to pass the locally instantiated provider directly to the operations that required it.  Good catch.